### PR TITLE
feat: self documenting test macro for layout tests ✨

### PR DESCRIPTION
**Pull request by kdheepak**
This macro

```rust
assert_size![
    let (area, flex) = (100, StretchLast),
    Length(25) => (x == 0, size == 25),
    Length(25) => (x == 25, size == 75),
];
```

expands to 

```rust
let area = 100;
let flex = Flex::StretchLast
let rect = Rect::new(0, 0, area, 1);
let constraints = [Constraint::Length(25), Constraint::Length(25)];
let intended_widths = [25, 75];
let intended_x = [0, 25];

let items = &Layout::horizontal().flex(flex).split(rect);
let widths: Vec<_> = items.iter().map(|item| item.width).collect();
assert_eq!(widths, intended_widths]);
let x: Vec<_> = items.iter().map(|item| item.x).collect();
assert_eq!(x, intended_x);
```

This makes it easier to read and tests and understand them without having to deal with whitespace or formatting.

### CODE REVIEW

**Review by Valentin271 — APPROVED** [review]
This otherwise looks good to me.

File: src/layout/layout.rs
Comment (kdheepak) [reply src/layout/layout.rs:779]: I added a function right below it showing the 4 ways you can use it. I can link to that test from the docstring and that should be sufficient I think?

Comment (Valentin271) [reply src/layout/layout.rs:779]: Right the testing function can act as documentation too. Yep that sufficient, just explicitly states that it also serves as doc.

Comment (kdheepak) [reply src/layout/layout.rs:781]: I've made an update adding docs.

**Comment by kdheepak** [comment]
Let's wait for @joshka's review before merging this since I know he had some opinions on how to do this.

**Comment by joshka** [comment]
I'd really suggest avoid making tests complicated like this. It looks neat, but it would be difficult to generally fix a test without intricate knowledge of the template macros. Let's instead find a way to get 90% of the same functionality with methods and super simple code.

Take this test:

```rust
        assert_size![
            // Proportional splits equally (widths only test)
            let (area, flex) = (100, Stretch),
            Length(25) => (x == 0, size == 50),
            Length(25) => (x == 50, size == 50),
        ];
```

We're testing that when you split an area of width 100, with stretch and two Length(25)s, then it returns two areas.
Something like this would keep the tests simple and obvious:

```rust
struct Split {
   start: u16,
   size: u16,
}

/// compute a list of splits for the size and constraint
fn splits(size: u16, flex: Flex, constraints: [Constraint]) -> Vec<Split> {
  let area = Rect::new(0, 0, size, 1);
  let items = &Layout::horizontal().flex(flex).split(rect);
  items.map(|item| SplitResult(start: item.x, size: item.width).collect();
}
```

This gives you simple tests, which are just as self describing as the onse above.
```rust
#[test]
fn proportional_splits_equally() {
  assert_eq!(
    splits(100, Stretch, [Length(25), Length(25)]);
    [Split { start: 0, size: 50 }, Split { start: 50, size: 50 }]
  }
);
```

I think the only thing missing here is that the constraint doesn't directly align with the start and position.

You might also do something like this:

```rust
fn splits<const N; usize>(size: u16, flex: Flex, constraints: [Constraint; N]) -> [Split; N] {
  let area = Rect::new(0, 0, size, 1);
  let layout = Layout::horizontal().flex(flex);
  let items = area.split(&layout);
  items.map(|item| SplitResult(start: item.x, size: item.width).collect();
}

#[test]
fn proportional_splits_equally() {
  let [length_1, length_2] = splits(100, Stretch, [Length(25), Length(25)]);
  assert_eq!(length_1, Split { start: 0, size: 50 });
  assert_eq!(length_2, Split { start: 50, size: 50 });
);
```

**Comment by kdheepak** [comment]
What do you think about this syntax:

```rust
assert_size![
    let (area, flex) = (100, Stretch),
    Length(25) => (start == 0, size == 50),
    Length(25) => (start == 50, size == 50),
];
```

but implemented as a one line `macro_rule` implementation that transforms to this:

```rust
assert_eq!(
    splits(100, Stretch, [Length(25), Length(25)]);
    [Split { start: 0, size: 50 }, Split { start: 50, size: 50 }]
}
```

That would make it easier to debug. 

Alternatively, we can leave it the way it is right now with this:

```rust
#[case::splits_equally_to_end(
    100, Flex::Stretch, 
    [Length(25), Length(25)], 
    [(0, 50), (50, 50)]
)]
```

---

What I'd really love to do is this:

```rust
assert_layout!(
    Flex::Stretch,
    <-----------------40 px---------------->
    ┌──────20 px───────┐┌──────20 px───────┐
    │      Min(20)     ││      Max(20)     │
    └──────────────────┘└──────────────────┘
)
```

It would be more verbose to type it out but it would be fully visually self describing.

**Comment by joshka** [comment]
I'd still say that adding a macro for testing should be a last resort. It's added complexity, which is something that always makes tests worse.

> That would make it easier to debug.

Show the errors that you get for both approaches. I'd bet the function approach has much easier to grok backtraces.

> What I'd really love to do is this:
```rust
assert_layout!(
    Flex::Stretch,
    <-----------------40 px---------------->
    ┌──────20 px───────┐┌──────20 px───────┐
    │      Min(20)     ││      Max(20)     │
    └──────────────────┘└──────────────────┘
)
```

I'd still want quotes around the layout, but that seems probably doable using functions still.
I think flipping the the constraint and the pixels would make sense.
Simplifying it to something like:
```rust
assert_eq!(
   render_split(Flex::Stretch, [Min(20), Max(20)]),
   buffer!("
        <--40px------>
        | Min(20 ||Max(20)|
    "));
```

This introduces the buffer macro, but does so in a way where it's more widely useful. Not 100% sure on this. Ideally you'd want the output to make it easy to paste back into the tests directly (which makes test generation quick and comparing failures quick), and this breaks down that.

**Pull request closed without merging**
